### PR TITLE
real egocentric for xworld2d; add navgation tasks for xworld3d; simplify c++ interface examples

### PR DESCRIPTION
**Pull request by yu239**
This is a mixture of three commits. See the title.

### CODE REVIEW

File: python/maze2d.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+            if not next in visited and next[0] >= 0 \
+               and next[0] < x and next[1] >= 0 and next[1] < y:
+                edges.add((cur, next))
+                dfs(next)
+
+    ## always start from (0, 0)
+    dfs((0, 0))
+    ## render the maze
+    for e in edges:
+        mid_x = e[0][0] + e[1][0]
```
Comment (Haichao-Zhang) [inline python/maze2d.py]: Why is mid_x and mid_y is calculated this way?  I can see x and y is half the size of X and Y, but each move still takes one unit.

File: python/py_init.cpp [hunk 2]
```diff
@@ -39,17 +40,15 @@ PyObject* get_gflag(PyObject* self, PyObject* args) {
...
-    } 
-    #ifdef XWORLD3D
-    else if (flag_name == "x3_task_mode") {
-        return PyString_FromString(FLAGS_x3_task_mode.c_str());
-    } else if (flag_name == "x3_reaching_distance") {
-        return Py_BuildValue("d", FLAGS_x3_reaching_distance);
-    }
-    #endif
-    else if (flag_name == "curriculum") {
+        //    } else if (flag_name == "x3_task_mode") {
```
Comment (Haichao-Zhang) [inline python/py_init.cpp]: why comment out x3_task_mode and x3_reaching_distance?

File: games/xworld3d/maps/xworld3d_env.py [hunk 3]
```diff
@@ -403,15 +315,14 @@ def __instantiate_entities(self):
         """
         Y, X = self.get_dims()
 
-        maze = self.__spanning_tree_maze_generator(X, Y)
+        maze = spanning_tree_maze_generator(X, Y)
         blocks = [(j, i, 0) for i,m in enumerate(maze) for j,b in enumerate(m) if b == '#']
-#        random.shuffle(blocks)
+        random.shuffle(blocks)
```
Comment (skylian) [inline games/xworld3d/maps/xworld3d_env.py]: add a comment to explain why shuffle, e.g. not all blocks will be used

File: games/xworld/xworld/xitem.h [hunk 4]
```diff
@@ -111,6 +122,8 @@ class XItem {
     // get the location of the item
     Loc get_item_location() { return Loc(int(e_.loc.x), int(e_.loc.y)); }
 
+    double get_item_yaw() { return e_.yaw; }
```
[the remark this answers is not in the document]
Comment (yu239) [reply games/xworld/xworld/xitem.h]: The icon can rotate 90, 180, and 270 degrees in the image plane.

File: games/xworld/xworld_simulator.cpp [hunk 5]
```diff
@@ -49,12 +48,14 @@ void XWorldSimulator::init() {
...
-        img_width_out_ = (width_ * 2 - 1) * block_size_;
+    if (FLAGS_visible_radius == 0) { // fully observed
+        int block_size = 12;
+        img_height_out_ = (height_ * 2 - 1) * block_size;
+        img_width_out_ = (width_ * 2 - 1) * block_size;
     } else {
-        img_height_out_ = height_ * block_size_;
-        img_width_out_ = width_ * block_size_;
+        int block_size = 28; // for partially observed, we can use larger grids
+        img_height_out_ = FLAGS_visible_radius * block_size;
```
[the remark this answers is not in the document]
Comment (yu239) [reply games/xworld/xworld_simulator.cpp]: That's possible. I should check the radius value.

File: python/maze2d.py (hunk 1 above)
Comment (yu239) [reply python/maze2d.py]: I first downsize the maze size by two. In the end, I need to multiply by two.

File: python/py_init.cpp (hunk 2 above)
Comment (yu239) [reply python/py_init.cpp]: They are no longer used.

File: games/xworld/maps/xworld_env.py [hunk 6]
```diff
@@ -76,25 +84,17 @@ def reset(self):
         self._configure()
         self.__instantiate_entities()
 
+    def get_current_usage(self):
+        usage = self.current_usage
+        self.current_usage = 0
```
[the remark this answers is not in the document]
Comment (yu239) [reply games/xworld/maps/xworld_env.py]: This part has been changed.

File: games/xworld/maps/xworld_env.py [hunk 7]
```diff
@@ -304,19 +307,46 @@ def __instantiate_entities(self):
...
+                e.loc = self.available_grids.pop()
+            ## if partially observed, perturb the objects
+            if get_flag("visible_radius") and e.type != "block":
+                ## four orientations
+                e.yaw = random.randint(-1,2) * self.PI_2
+                if e.type == "goal":
+                    ## random scale
+                    e.scale = random.uniform(0.5, 1)
+                    ## random offset
+                    e.offset = random.uniform(0, 1 - e.scale)
```
[the remark this answers is not in the document]
Comment (yu239) [reply games/xworld/maps/xworld_env.py]: Currently the top left corner is always on the diagonal line of a grid.

File: games/xworld/xworld/xitem.cpp [hunk 8]
```diff
@@ -40,10 +42,91 @@ cv::Mat XItem::get_item_image() {
                        cv::INTER_LINEAR);
             item_imgs_[img_name] = img;
         }
-        return item_imgs_[img_name];
+
+        // transform the icon
+        auto icon = item_imgs_[img_name].clone();
+        cv::Point2f center(icon.cols / 2.0, icon.rows / 2.0);
+        auto rot_mat = cv::getRotationMatrix2D(center, e_.yaw * 180 / M_PI, e_.scale);
+        rot_mat.at<double>(0, 2) += (e_.offset + e_.scale / 2 - 0.5) * icon.cols;
```
[the remark this answers is not in the document]
Comment (yu239) [reply games/xworld/xworld/xitem.cpp]: e_.offset + e_.scale / 2 - 0.5 computes the translation of the icon image inside the grid

**Pull request merged**